packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts

```diff
@@ -577,6 +613,16 @@ export class Pipeline extends PipelineBase {
 
   private getRoleFromActionPropsOrGenerateIfCrossAccount(stage: Stage, action: IAction): iam.IRole | undefined {
     const pipelineStack = Stack.of(this);
+    const actionEnv = new RichAction(action).env;
+
+    // Do a check on whether the cross-account acccess is possible given the
+    // current pipeline configuration. This has bearing on whether
+    // `artifactBucket.grantRead(action.role)` (which is going to occur down the
+    // line) can even work.
+    const hasAccount = !Token.isUnresolved(actionEnv.account);
+    if (hasAccount && !sameEnvDimension(this.env.account, actionEnv.account) && !this.crossAccountKeys) {
```

Comment:
One thing: I see we already do a check for this [here](https://github.com/aws/aws-cdk/blob/e349004a522e2123c1e93bd3402dd7c3f9c5c17c/packages/%40aws-cdk/aws-codepipeline/lib/pipeline.ts#L615-L618), and it's more general, because it checks for the case when a customer provides their own Bucket (which this does not).

Do we really need this extra check here? Is there something missing from the existing check that we should change?

Reply:
You're right, this is way better.

Reply:
Aha it's actually not. If I just rely on that check, the tests for "cross-account by role" and "cross-account by resource" stop throwing the errror.

For the first case because if we use a Role we never get there: https://github.com/aws/aws-cdk/blob/huijbers/cheap-pipelines/packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts/#L642

The second case I haven't figured out yet.

But in fact this is also wrong, if we were to be COMPLETELY correct we should not be checking the `this.artifactBucket`, we should be checking the bucket in the same region -- which *can* be the `artifactBucket` but it can also be one of the `replicationBucket`s.

Which in fact suggests that the method that deals with the region is the better place to put this check?

Reply:
That's all fine by me, let's correct this check instead of adding a new one - we have to tackle the case of a customer passing a Bucket without a Key anyway, might as well do it immediately. I can't imagine getting the correct Bucket (replication or main) for an action's region in that method that deals with accounts will be difficult, and I think having a nice separation between cross-region setup done in `ensureReplicationResourcesExistFor` and cross-account setup done in `getRoleForAction` is important (the code here is already complicated and difficult to follow).

Reply:
No but it's silly, because the only function that properly returns the bucket is `ensureReplicationResourcesExistFor`, so that's the one we're going to have to use.

So now this is the call tree:

```
_attachActionToPipeline
  -> ensureReplicationResourcesExistFor (to get cross-region info)
  -> getRoleForAction (to get the role)
        -> ensureReplicationResourcesExistFor (to get the cross-region bucket and check if it has a key)
```

So I'm going to do it, but I strongly feel this is silly.